stubs/10.0.0/HasManyThrough.stub

```diff
@@ -48,6 +48,34 @@ class HasManyThrough extends Relation
      */
     public function cursorPaginate($perPage = null, $columns = ['*'], $cursorName = 'cursor', $cursor = null);
 
+    /**
+     * Execute the query and get the first related model.
+     *
+     * @param  array<int, string>  $columns
+     * @return TRelatedModel|null
+     */
+    public function first($columns = ['*']);
+
+    /**
+     * Execute the query and get the first result or throw an exception.
+     *
+     * @param  array<int, string>  $columns
+     * @return TRelatedModel
+     *
+     * @throws \Illuminate\Database\Eloquent\ModelNotFoundException<TRelatedModel>
+     */
+    public function firstOrFail($columns = ['*']);
+
+    /**
+     * Execute the query and get the first result or call a callback.
+     *
+     * @template TReturnClosure
+     * @param  \Closure(): TReturnClosure|array<int, string>  $columns
+     * @param  \Closure(): TReturnClosure|null  $callback
+     * @return TRelatedModel|TReturnClosure
+     */
```

Comment:
This need to be:

```suggestion
    /**
     * Execute the query and get the first result or call a callback.
     *
     * @template TFirstOrValue
     *
     * @param  \Closure|array  $columns
     * @param  (\Closure(): TFirstOrValue)|null  $callback
     * @return ($callback is null ? TRelatedModel|null : TRelatedModel|TFirstOrValue)
     */
```

Reply:
Thanks for the suggestion, I originally had exactly that, but I think this is incorrect.
If you look at the implementation, it is implicitly required to provide a callback. The only reason the second argument is nullable is that it is possible to provide a callback in the first argument. If you would run `->firstOr()` and the query return no result you would get an exception not `null`, since `$callback` is `null` and `null` is not callable.

```php
public function firstOr($columns = ['*'], ?Closure $callback = null)
{
    if ($columns instanceof Closure) {
        $callback = $columns;

        $columns = ['*'];
    }

    if (! is_null($model = $this->first($columns))) {
        return $model;
    }

    return $callback();
}
```